substrate/frame/multisig-stateful/src/lib.rs

```diff
@@ -0,0 +1,1019 @@
+				Error::<T>::NotEnoughApprovers
+			);
+
+			Self::check_approvers_still_exist(
+				&multisig_account,
+				&multisig_details,
+				&multisig_proposal,
+				&call_hash,
+			)?;
+
+			// Remove the proposal from the state.
+			PendingProposals::<T>::remove(&multisig_account, &call_hash);
+
+			// Make the call the last thing to prevent any possible re-entrancy attacks
+			let result = call
+				.dispatch(RawOrigin::Signed(multisig_account.clone()).into())
+				.map(|_| ())
+				.map_err(|e| e.error);
+
+			Self::return_proposal_deposit(&multisig_proposal)?;
+
+			Self::deposit_event(Event::ExecutedProposal {
+				executor: who,
+				multisig_account,
+				call_hash,
+				result,
+			});
+
+			result
+		}
+
+		/// Cancels an existing proposal for a multisig account Only if the proposal doesn't have approvers other than
+		/// the proposer.
+		///
+		///	This function needs to be called from a the proposer of the proposal as the origin.
+		///
+		/// The withheld deposit will be returned to the proposal creator.
+		///
+		/// # Arguments
+		///
+		/// * `multisig_account` - The multisig account ID.
+		/// * `call_hash` - The hash of the call to be canceled. (This will be the hash of the call that was used in `start_proposal`)
+		///
+		/// # Errors
+		///
+		/// * `MultisigNotFound` - The multisig account does not exist.
+		/// * `ProposalNotFound` - The proposal does not exist.
+		#[pallet::call_index(5)]
+		#[pallet::weight(Weight::default())]
+		pub fn cancel_own_proposal(
+			origin: OriginFor<T>,
+			multisig_account: T::AccountId,
+			call_hash: T::Hash,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(
+				MultisigAccount::<T>::contains_key(&multisig_account),
+				Error::<T>::MultisigNotFound
+			);
+			let multisig_proposal = PendingProposals::<T>::get(&multisig_account, &call_hash)
+				.ok_or(Error::<T>::ProposalNotFound)?;
+
+			ensure!(
+				multisig_proposal.approvers.len() == 1
+					&& multisig_proposal.approvers.contains(&who),
+				Error::<T>::UnAuthorizedOwner
+			);
+
+			PendingProposals::<T>::remove(&multisig_account, &call_hash);
+			Self::return_proposal_deposit(&multisig_proposal)?;
+			Self::deposit_event(Event::CanceledProposal { multisig_account, call_hash });
+			Ok(())
+		}
+
+		/// Remove up to `max` stale proposals for a deleted multisig account.
+		///
+		/// The withheld deposit will be returned to each proposal creator.
+		///
+		/// May be called by any Signed origin, but only after the multisig account is deleted.
+		#[pallet::call_index(25)]
+		#[pallet::weight(Weight::default())]
+		pub fn cleanup_proposals(
+			origin: OriginFor<T>,
+			multisig_account: T::AccountId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+			ensure!(
+				!MultisigAccount::<T>::contains_key(&multisig_account),
+				Error::<T>::MultisigStillExists
+			);
+
+			// Although we're iterating over dreaind proposals, It's hard to attack the runtime by having a very large
```

Comment:
This is an unbounded iteration over all the proposals. What is the best way of achieving an iteration but limited? I need to get the proposals themselves to be able to return the deposit. To my knowledge `clear_prefix` with a limit doesn't get the proposals.